src/resolver.rs

```diff
@@ -2161,6 +2166,131 @@ impl FailForViolationConflict {
+        from_target: SortedSet<&'a Option<VetVersion>>,
+    }
+    let mut reachable = None::<Reachable<'_>>;
+    for SearchFailure {
+        reachable_from_root,
+        reachable_from_target,
+    } in failures
+    {
+        if let Some(Reachable {
+            from_root,
+            from_target,
+        }) = reachable.as_mut()
+        {
+            // FIXME: this is horrible but I'm tired and this avoids false-positives
+            // and duplicates. This does the right thing in the common cases, by
+            // restricting ourselves to the reachable nodes that are common to all
+            // failures, so that we can suggest just one change that will fix
+            // everything.
+            from_root.retain(|ver| reachable_from_root.contains(ver));
+            from_target.retain(|ver| reachable_from_target.contains(ver));
+        } else {
+            // We only have sources available for the
+            // package itself, and any non-git versions.
+            let version_has_sources = |ver: &&Option<VetVersion>| -> bool {
+                ver.as_ref() == Some(&package.version)
+                    || !matches!(
+                        ver,
+                        Some(VetVersion {
+                            git_rev: Some(_),
+                            ..
+                        })
+                    )
+            };
+            reachable = Some(Reachable {
+                from_root: reachable_from_root
+                    .iter()
+                    .filter(version_has_sources)
+                    .collect(),
+                from_target: reachable_from_target
+                    .iter()
+                    .filter(version_has_sources)
+                    .collect(),
+            });
+        }
+    }
+
+    let Some(Reachable { from_root, from_target }) = &reachable else {
+        // Nothing failed, return a dummy suggestion for an empty diff.
+        return Some(DiffRecommendation {
+            from: Some(package.version.clone()),
+            to: package.version.clone(),
+            diffstat: DiffStat { insertions: 0, deletions: 0, files_changed: 0 },
+        });
+    };
+
+    // Now suggest solutions of those failures
+    let mut candidates = SortedSet::new();
+    for &dest in from_target {
+        let mut closest_above = None;
+        let mut closest_below = None;
+        for &src in from_root {
+            if src < dest {
+                if let Some(closest) = closest_below {
+                    if src > closest {
+                        closest_below = Some(src);
+                    }
+                } else {
+                    closest_below = Some(src);
+                }
+            } else if let Some(closest) = closest_above {
+                if src < closest {
+                    closest_above = Some(src);
+                }
+            } else {
+                closest_above = Some(src);
+            }
+        }
```

Comment:
I realize it'd be doubling the iterations, but I think this would be a lot more readable as:
```suggestion
        let closest_above = from_root.iter().filter(|src| src >= dest).min_by_key(|src| src - dest);
        let closest_below = from_root.iter().filter(|src| src < dest).min_by_key(|src| dest - src);
```

Reply:
It actually occurs to me that `BTreeSet` is sorted already, so we can use the range syntax, and probably write it like:

```rust
        let closest_above = from_root.range(dest..).next();
        let closest_below = from_root.range(..dest).next_back();
```

Though it seems like this will require a turbofish to make it work with inference because of the reference - not the end of the world.

Reply:
:facepalm: haha yeah that's much better. I didn't register they were `SortedSet`!